_source/_posts/2021-12-03-angular-dynamic-components.md

```diff
@@ -0,0 +1,773 @@
+  `
+})
+export class MenuComponent implements OnInit, OnDestroy {  
+  public isAuthenticated = false;  
+  private _destroySub$ = new Subject<void>();  
+  
+  constructor(private _oktaAuth: OktaAuth, private _authStateService: OktaAuthStateService, private _router: Router) { }  
+  
+  public ngOnInit(): void {  
+   this._authStateService.authState$.pipe(  
+     filter((s: AuthState) => !!s),  
+     map((s: AuthState) => s.isAuthenticated ?? false),  
+     distinctUntilChanged(),  
+     takeUntil(this._destroySub$)
+   ).subscribe(
+     (authenticated: boolean) => this.isAuthenticated = authenticated
+   );  
+  }  
+
+  public ngOnDestroy(): void {
+    this._destroySub.next();
+  }
+  
+  public async login(): Promise<void> {  
+    await this._oktaAuth.signInWithRedirect().then(  
+      _ => this._router.navigate(['/protected'])  
+    );  
+  }  
+  
+  public async logout(): Promise<void> {  
+    await this._oktaAuth.signOut();  
+  }  
+}
+```
+{% endraw %}
+
+Now that you can log in let's get your name displayed in the toolbar. Open `src/app/profile/profile.component.ts`.  Okta's auth state has user info. Notice it's also available through a claim. Replace the `Profile` component code with the following.
+
+{% raw %}
+```ts
+@Component({  
+  selector: 'app-profile',  
+  template: `  
+    <ng-container *ngIf="name$ | async as name ">
+      <span class="mat-body-1">{{name}}</span>  
+    </ng-container> 
+`})  
+export class ProfileComponent {  
+  public name$: Observable<string> = this._authStateService.authState$.pipe( 
+    filter((s: AuthState) => !!s && !!s.isAuthenticated),  
+    map((s: AuthState) => s.idToken?.claims.name ?? '')  
+  );  
+  
+  constructor(private _authStateService: OktaAuthStateService) { }  
+}
+```
+{% endraw %}
+
+Now you see your name when you log in to the application.
+
+## Format the dynamic components
+
+All the messages have data to display. The data always includes an URL plus other content that is unique to each component. Let's define the interface to support the message data. Run the following command to create the interface.
+
+```bash
+ng generate interface message
+```
+
+There'll be three different message formats, one for each dynamic component. So we'll define a message type, standardized data properties, and a class that holds the type and data properties. Open `src/app/message.ts` and replace the contents with the following code.
+
+```ts
+export type MessageType = 'Pawesome' | 'Clawesome' | 'Smiley';  
+  
+export interface MessageData {  
+  url: string;  
+  content?: any;  
+}  
+  
+export class MessageItem {  
+  constructor(public type: MessageType, public data: MessageData) { }  
+}
+```
+
+When we created the dynamic components earlier, they all implemented the same base interface, `DynamicComponent`. Since all the dynamic components have some data, we need to update the `DynamicComponent` interface to reflect this shared property that all components will implement.
```

Comment:
```suggestion
When we created the dynamic components earlier, they all implemented the same base interface, `DynamicComponent`. Since all the dynamic components have some data, we need to update the `DynamicComponent` interface to reflect this shared property that all the components will implement.
```